### PR TITLE
[aggregator] Improve stream estimation

### PR DESCRIPTION
**Pull request by vdarulis**
**What this PR does / why we need it**:

Fix a bunch of issues, including and not limited to:
* alloc hell
* having effective `insertAndCompressEvery` == 0, so we end up compressing every `flushEvery` interval
* insert being completely broken, inserting only $numelements * $eps every call - a translation error from original statiste code, since it did not have an explicit value for when to do the insert+compress cycle
* options not working, if not overriding the stream/float pools

Despite deprecation of a bunch of tunables, the code performs better on average w/o micromanagement of pool sizes, with far lower memory consumption for heavy timers, as we no longer excessively buffer sample values in the heap. Eventually, we should switch to a more modern quantile estimation algo, something like KLL, that has better properties.

Note that one of benchmarks doesn't complete on the old code, as it exceeds the error threshold (probably due to broken insert logic), but the rest of the results are fine:
```
name                                                              old time/op    new time/op    delta
TimerValues-12                                                      17.6µs ± 3%     0.0µs ± 1%   -99.80%  (p=0.008 n=5+5)
TimerValueOf-12                                                     79.4ns ± 1%    74.4ns ± 0%    -6.32%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size-12                       4.54ms ±20%    1.54ms ±13%   -66.08%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..1,000,000_range-12    4.39ms ± 4%    1.63ms ± 4%   -62.77%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..100,000_range-12      4.45ms ± 2%    1.69ms ± 3%   -62.13%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_1000_batch_size-12                         414µs ± 7%     205µs ± 2%   -50.57%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size-12                          421µs ± 7%     201µs ± 3%   -52.20%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size_with_negative_values-12     425µs ± 9%     199µs ± 2%   -53.05%  (p=0.008 n=5+5)

name                                                              old alloc/op   new alloc/op   delta
TimerValues-12                                                       1.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
TimerValueOf-12                                                      0.00B          0.00B           ~     (all equal)
TimerAddBatch/100k_samples_1000_batch_size-12                       6.90MB ± 0%    0.00MB ±18%   -99.98%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..1,000,000_range-12    6.90MB ± 0%    0.00MB ±24%   -99.98%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..100,000_range-12      6.90MB ± 0%    0.00MB ± 3%   -99.98%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_1000_batch_size-12                         743kB ± 0%       1kB ± 2%   -99.90%  (p=0.016 n=5+4)
TimerAddBatch/10k_samples_100_batch_size-12                          743kB ± 0%       1kB ± 3%   -99.90%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size_with_negative_values-12     743kB ± 0%       1kB ± 4%   -99.92%  (p=0.008 n=5+5)

name                                                              old allocs/op  new allocs/op  delta
TimerValues-12                                                        0.00           0.00           ~     (all equal)
TimerValueOf-12                                                       0.00           0.00           ~     (all equal)
TimerAddBatch/100k_samples_1000_batch_size-12                         100k ± 0%        0k ± 0%   -99.99%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..1,000,000_range-12      100k ± 0%        0k ± 0%   -99.99%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..100,000_range-12        100k ± 0%        0k ± 0%   -99.99%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_1000_batch_size-12                         10.0k ± 0%      0.0k ± 0%   -99.88%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size-12                          10.0k ± 0%      0.0k ± 0%   -99.88%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size_with_negative_values-12     10.0k ± 0%      0.0k ± 0%   -99.88%  (p=0.008 n=5+5)

name                                                              old speed      new speed      delta
TimerAddBatch/100k_samples_1000_batch_size-12                      178MB/s ±17%   524MB/s ±12%  +193.66%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..1,000,000_range-12   182MB/s ± 4%   490MB/s ± 4%  +168.53%  (p=0.008 n=5+5)
TimerAddBatch/100k_samples_1000_batch_size_0..100,000_range-12     180MB/s ± 2%   474MB/s ± 3%  +164.08%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_1000_batch_size-12                       193MB/s ± 7%   391MB/s ± 2%  +102.06%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size-12                        190MB/s ± 7%   398MB/s ± 3%  +108.84%  (p=0.008 n=5+5)
TimerAddBatch/10k_samples_100_batch_size_with_negative_values-12   189MB/s ± 9%   401MB/s ± 2%  +112.47%  (p=0.008 n=5+5)
```
**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/aggregator/aggregation/quantile/cm/heap.go [hunk 1]
```diff
@@ -34,50 +34,66 @@ func (h minHeap) Min() float64 { return h[0] }
...
 		parent := (i - 1) / 2
-		if parent == i || h[parent] <= h[i] {
+		if parent == i || heap[parent] <= heap[i] {
 			break
 		}
-		h[parent], h[i] = h[i], h[parent]
+		heap[parent], heap[i] = heap[i], heap[parent]
 		i = parent
 	}
 }
```
Comment (abliqo) [inline src/aggregator/aggregation/quantile/cm/heap.go:54]: Doesn't `*h` need to be updated with modified `heap` in the end?

File: src/aggregator/aggregation/quantile/cm/heap.go [hunk 2]
```diff
@@ -34,50 +34,66 @@ func (h minHeap) Min() float64 { return h[0] }
...
 
-func (h minHeap) heapify(i, n int) {
-	for {
-		left := i*2 + 1
+func (h *minHeap) ensureSize() {
+	heap := *h
+	targetCap := cap(heap) * 2
+
+	newHeap := sharedHeapPool.Get(targetCap)
+	(*newHeap) = append(*newHeap, heap...)
```
Comment (abliqo) [inline src/aggregator/aggregation/quantile/cm/heap.go:61]: Doesn't that result in `newHeap` having two sections?
- first (bigger) section with blank values 
- second section with existing values

Comment (vdarulis) [reply src/aggregator/aggregation/quantile/cm/heap.go:61]: pool resets the slice len to 0 (but it has the right cap):
```

		*v = (*v)[:0]
		return v
	}
makeNew:
	newVal := make(minHeap, 0, size)

	return &newVal
```

File: src/aggregator/aggregation/quantile/cm/heap.go (hunk 1 above)
Comment (vdarulis) [reply src/aggregator/aggregation/quantile/cm/heap.go:54]: nope, it's just a swap that mutates slice members. 
we only need to update the original reference due to reslice (append) operations.

**Review by abliqo — APPROVED** [review]
LGTM

**Pull request merged**